### PR TITLE
[CARBONDATA-3511] Query time improvement by reducing the number of NameNode calls while having carbonindex files in the store

### PR DESCRIPTION
**Pull request by dhatchayani**
**Problem:**
If the store contains only carbonindex files, while querying that table, the index file to block details mapping needs to be formed. This formation is currently segment level. So the map formed will be of SegmentPath to all its indexFilePath to its block details. This is not reused and the same map is formed for each and every indexFile. This increases the name node calls as for each indexFile it needs to form the map segment level(which includes all the indexFiles).

**Solution:**
Make the map global and reuse it in needed times and clear the same in needed scenarios.

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [x] Testing done
        No.of Index Files = 110
        Query Time in seconds

| **Before Changes** |  **After changes**|
|--|--|
|16.665  | 2.077 |
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/403/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/403/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/402/

File: core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java [hunk 1]
```diff
@@ -234,6 +231,10 @@ private BlockletDataMapIndexWrapper get(TableBlockIndexUniqueIdentifierWrapper i
                 tableSegmentUniqueIdentifierWrapper.isAddTableBlockToUnsafeAndLRUCache());
       }
     }
+    String segmentPath =
+        tableSegmentUniqueIdentifierWrapper.getTableBlockIndexUniqueIdentifier().getIndexFilePath();
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java]: `getIndexFilePath()` returns `segmentPath`?

File: core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java [hunk 2]
```diff
@@ -64,6 +64,13 @@
    */
   private Map<String, Object> segmentLockMap;
 
+  /**
+   * map of segmentPath to the map of index/ merge index file path to its
+   * corresponding block info. While querying the table this map will be useful while creating the
+   * block info for the index or merge index file.
+   */
+  private Map<String, Map<String, BlockMetaInfo>> segInfoCache;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java]: Global cache will be risky, how to clean it? Better have a method level cache.

File: core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java (hunk 1 above)
Comment (dhatchayani) [reply core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java]: yes. getIndexFilePath() will return the indexFile parent path, which is segmentPath

File: core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java (hunk 2 above)
Comment (dhatchayani) [reply core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java]: ok.. we can reuse the getAll method to reuse the method level cache.
Closing this PR

**Comment by dhatchayani** [comment]
As Global level cache is risky and need to be cleaned at needed scenarios, we can reuse the method level cache from the getAll method. Closing this PR

**Pull request closed without merging**